### PR TITLE
FX-only codegen

### PR DESCRIPTION
**Pull request by Linux-cpp-lisp**
## Description
Initial work on FX-only codegen:

- [X] Pure FX TensorProduct
- [X] Pure FX Linear
- [X] Pure FX Extract / ExtractIr
- [X] Simplified `ExtractIr` without submodules
- [X] `ReducedTensorProduct` without `Ellipsis` using [`torch.narrow`](https://pytorch.org/docs/stable/generated/torch.narrow.html) — this removes the need for PyTorch 1.8.1, allowing 1.8.0 support to be maintained

Issues remaining:

- [x] Use of `eval_code` in `angular_spherical_harmonics` has to be somehow changed if we want to remove `eval_code` entirely
- [x] Copying + saving of models: pickling a `ScriptModule` isn't possible, so `TensorProduct` and `Linear` may still need to use a modified version of `CodeGenMixin` to manage pickle state with `fx.Graph`s instead of strings. Not clear yet what has to be done here.

## How Has This Been Tested?
e3nn test suite

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/e3nn/e3nn/blob/main/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [X] I have updated the documentation (if relevant).
- [X] I have added tests that cover my changes (if relevant).
- [ ] All new and existing tests passed.
- [ ] I have updated the [Changelog](https://github.com/e3nn/e3nn/blob/main/ChangeLog.md).

### CODE REVIEW

File: e3nn/nn/_extract.py [hunk 1]
```diff
@@ -37,43 +43,41 @@ def __init__(self, irreps_in, irreps_outs, instructions):
...
-            "@torch.jit.script",
-            "def main(x: torch.Tensor):",
-            f"    assert x.shape[-1] == {self.irreps_in.dim}",
-            "    out = (",
-        ]
-        s = " "*4
+        # == generate code ==
+        graph = self.graph
+        x = fx.Proxy(graph.placeholder('x', torch.Tensor))
+        torch._assert(x.shape[-1] == self.irreps_in.dim, "invalid input shape")
```
Comment (mariogeiger) [inline e3nn/nn/_extract.py:49]: Nice!

File: e3nn/nn/_extract.py [hunk 2]
```diff
@@ -85,13 +89,9 @@ def __init__(self, irreps_in, ir):
...
-        instructions = [tuple(i for i, mul_ir in enumerate(self.irreps_in) if mul_ir.ir == ir)]
-
-        self.ext = Extract(self.irreps_in, [self.irreps_out], instructions)
+        irreps_in = o3.Irreps(irreps_in)
+        self.irreps_out = o3.Irreps([mul_ir for mul_ir in irreps_in if mul_ir.ir == ir])
+        instructions = [tuple(i for i, mul_ir in enumerate(irreps_in) if mul_ir.ir == ir)]
 
-    def forward(self, x):
-        return self.ext(x)[0]
+        super().__init__(irreps_in, [self.irreps_out], instructions, squeeze_out=True)
```
Comment (mariogeiger) [inline e3nn/nn/_extract.py:97]: Clean!

**Comment by Linux-cpp-lisp** [comment]
I was hoping to save TorchScript generated code as TorchScript IR, but unfortunately it does not appear to be possible due to this issue: https://github.com/pytorch/pytorch/issues/27343

**Comment by Linux-cpp-lisp** [comment]
Resolved #249

**Pull request merged**